File: blockchain-parser/src/withdrawalService.js
const { ethers } = require('ethers');
const config = require('./config');
const { AppDataSource } = require('./database');
const Account = require('./entity/Account');
const Withdrawal = require('./entity/Withdrawal');
const { ERC20_ABI } = require('./blockchainParser');


class WithdrawalService {
  constructor() {
    this.provider = new ethers.JsonRpcProvider(config.blockchain.rpcUrl);
    this.wallet = new ethers.Wallet(process.env.SEPOLIA_PRIVATE_KEY, this.provider);
    this.usdtContract = new ethers.Contract(config.blockchain.usdtAddress, ERC20_ABI, this.wallet);
  }

  // 处理提现请求
  async processWithdrawals() {
    const withdrawalRepo = AppDataSource.getRepository(Withdrawal);
    const pendingWithdrawals = await withdrawalRepo.find({
      where: { status: 'init' },
      relations: ['account']
    });

    for (const withdrawal of pendingWithdrawals) {
      // 更新状态为处理中
      if (withdrawal.status !== 'init') {
        continue;
      }

      withdrawal.status = 'processing';
      await withdrawalRepo.save(withdrawal);
    }

    for (const withdrawal of pendingWithdrawals) {
      await this.processWithdrawal(withdrawal);
    }
  }

  // 处理单个提现
  async processWithdrawal(withdrawal) {
    const withdrawalRepo = AppDataSource.getRepository(Withdrawal);
    const accountRepo = AppDataSource.getRepository(Account);

    try {
      if (withdrawal.status !== 'processing') {
        return;
      }

      console.log(`Processing withdrawal ${withdrawal.id}...`);

      let txHash;

      // BNB 提现
      if (withdrawal.token_symbol === 'BNB') {
        txHash = await this.withdrawBNB(withdrawal);
      }
      // USDT 提现
      else if (withdrawal.token_symbol === 'USDT') {
        txHash = await this.withdrawUSDT(withdrawal);
      }

      // 更新提现记录
      withdrawal.tx_hash = txHash;
      await withdrawalRepo.save(withdrawal);

      // 监听交易结果
      await this.monitorTransaction(txHash, withdrawal);

    } catch (error) {
      console.error(`Withdrawal ${withdrawal.id} failed:`, error);
      withdrawal.status = 'failed';
      await withdrawalRepo.save(withdrawal);
    }
  }

  // 提现 BNB
  async withdrawBNB(withdrawal) {
    const tx = await this.wallet.sendTransaction({
      to: withdrawal.to_address,
      value: ethers.parseEther(withdrawal.amount.toString()),
      gasLimit: 21000
    });

    return tx.hash;
  }

  // 提现 USDT
  async withdrawUSDT(withdrawal) {
    const decimals = await this.usdtContract.decimals();
    const amount = ethers.parseUnits(withdrawal.amount.toString(), decimals);
    const tx = await this.usdtContract.transfer(withdrawal.to_address, amount);
    return tx.hash;
  }

  // 监听交易结果
  async monitorTransaction(txHash, withdrawal) {
    const withdrawalRepo = AppDataSource.getRepository(Withdrawal);
    const accountRepo = AppDataSource.getRepository(Account);
    
    try {
      const receipt = await this.provider.waitForTransaction(txHash, 3); // 等待3个确认
      
      if (receipt.status === 1) {
        // 提现成功
        withdrawal.status = 'success';
        await withdrawalRepo.save(withdrawal);
        
        // 这里不需要更新用户余额，因为提现时已经扣减
        console.log(`Withdrawal ${withdrawal.id} succeeded`);
      } else {
        // 提现失败，回滚用户余额
        withdrawal.status = 'failed';
        await withdrawalRepo.save(withdrawal);
        
        // 回滚用户余额
        if (withdrawal.token_symbol === 'BNB') {
          await accountRepo.increment(
            { id: withdrawal.account_id },
            'bnb_amount',
            withdrawal.amount
          );
        } else if (withdrawal.token_symbol === 'USDT') {
          await accountRepo.increment(
            { id: withdrawal.account_id },
            'usdt_amount',
            withdrawal.amount
          );
        }
        
        console.log(`Withdrawal ${withdrawal.id} failed, rolled back balance`);
      }
    } catch (error) {
      console.error(`Error monitoring transaction ${txHash}:`, error);
      withdrawal.status = 'failed';
      await withdrawalRepo.save(withdrawal);
    }
  }
}

module.exports = WithdrawalService;

File: blockchain-parser-go/service/withdrawal_service.go
package service

import (
	"bytes"
	"context"
	"fmt"
	"log"
	"math/big"
	"os"
	"time"

	"blockchain-parser-go/config"
	"blockchain-parser-go/database"
	blockchainTypes "blockchain-parser-go/types"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/accounts/abi/bind"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
)

type WithdrawalService struct {
	client       *ethclient.Client
	wallet       *bind.TransactOpts
	usdtAddress  common.Address
	usdtABI      abi.ABI
	db           *database.DB
	scanInterval time.Duration
}

func NewWithdrawalService(cfg *config.Config, db *database.DB) (*WithdrawalService, error) {
	client, err := ethclient.Dial(cfg.RPCURL)
	if err != nil {
		return nil, err
	}

	privateKey, err := crypto.HexToECDSA(cfg.PrivateKey)
	if err != nil {
		return nil, err
	}

	chainID := big.NewInt(cfg.ChainID)
	wallet, err := bind.NewKeyedTransactorWithChainID(privateKey, chainID)
	if err != nil {
		return nil, err
	}

	myERC20ABIData, err := os.ReadFile("abis/MyERC20.json")
	if err != nil {
		return nil, fmt.Errorf("failed to read ABI: %w", err)
	}

	usdtABI, err := abi.JSON(bytes.NewReader(myERC20ABIData))
	if err != nil {
		return nil, fmt.Errorf("failed to parse ABI: %w", err)
	}

	return &WithdrawalService{
		client:       client,
		wallet:       wallet,
		usdtAddress:  common.HexToAddress(cfg.USDTAddress),
		usdtABI:      usdtABI,
		db:           db,
		scanInterval: time.Duration(cfg.ScanInterval) * time.Millisecond,
	}, nil
}

func (s *WithdrawalService) Start() {
	go s.processWithdrawals()
	log.Println("Withdrawal service started")
}

func (s *WithdrawalService) processWithdrawals() {
	ticker := time.NewTicker(s.scanInterval)
	defer ticker.Stop()

	for range ticker.C {
		withdrawals, err := s.db.GetPendingWithdrawals()
		if err != nil {
			log.Printf("Error getting pending withdrawals: %v", err)
			continue
		}

		for _, withdrawal := range withdrawals {
			if err := s.processWithdrawal(withdrawal); err != nil {
				log.Printf("Error processing withdrawal %d: %v", withdrawal.ID, err)
			}
		}
	}
}

func (s *WithdrawalService) processWithdrawal(withdrawal *blockchainTypes.Withdrawal) error {
	// Update status to processing
	if err := s.db.UpdateWithdrawalStatus(withdrawal.ID, "processing", ""); err != nil {
		return err
	}

	var tx *types.Transaction
	var err error

	switch withdrawal.TokenSymbol {
	case "BNB":
		tx, err = s.withdrawBNB(withdrawal)
	case "USDT":
		tx, err = s.withdrawUSDT(withdrawal)
	default:
		err = fmt.Errorf("unsupported token symbol: %s", withdrawal.TokenSymbol)
	}

	if err != nil {
		s.db.UpdateWithdrawalStatus(withdrawal.ID, "failed", "")
		return err
	}

	// Update with transaction hash
	if err := s.db.UpdateWithdrawalStatus(withdrawal.ID, "processing", tx.Hash().Hex()); err != nil {
		return err
	}

	// Monitor transaction
	go s.monitorTransaction(tx, withdrawal)

	return nil
}

func (s *WithdrawalService) withdrawBNB(withdrawal *blockchainTypes.Withdrawal) (*types.Transaction, error) {
	nonce, err := s.client.PendingNonceAt(context.Background(), s.wallet.From)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %w", err)
	}

	// 获取当前 gas price
	gasPrice, err := s.client.SuggestGasPrice(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %w", err)
	}

	tx := types.NewTransaction(
		nonce,
		common.HexToAddress(withdrawal.ToAddress),
		withdrawal.Amount,
		21000,
		gasPrice,
		nil,
	)

	signedTx, err := s.wallet.Signer(s.wallet.From, tx)
	if err != nil {
		return nil, err
	}

	if err := s.client.SendTransaction(context.Background(), signedTx); err != nil {
		log.Printf("[DEBUG] Failed to send transaction: %v", err)
		return nil, err
	}

	return signedTx, nil
}

func (s *WithdrawalService) withdrawUSDT(withdrawal *blockchainTypes.Withdrawal) (*types.Transaction, error) {
	// For USDT, we need to use the contract's transfer function
	// This is a simplified version - in practice you'd use the contract ABI
	// to properly encode the transfer function call
	nonce, err := s.client.PendingNonceAt(context.Background(), s.wallet.From)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %w", err)
	}

	// 获取当前 gas price
	gasPrice, err := s.client.SuggestGasPrice(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %w", err)
	}

	tx := types.NewTransaction(
		nonce,
		s.usdtAddress,
		big.NewInt(0), // ERC20 转账不需要发送原生代币（BNB），value 设为 0
		100000,
		gasPrice,
		s.encodeTransferCall(withdrawal.ToAddress, withdrawal.Amount),
	)

	signedTx, err := s.wallet.Signer(s.wallet.From, tx)
	if err != nil {
		return nil, err
	}

	if err := s.client.SendTransaction(context.Background(), signedTx); err != nil {
		log.Printf("[DEBUG] Failed to send transaction: %v", err)
		return nil, err
	}

	return signedTx, nil
}

func (s *WithdrawalService) monitorTransaction(tx *types.Transaction, withdrawal *blockchainTypes.Withdrawal) {
	// Wait for transaction to be mined
	receipt, err := bind.WaitMined(context.Background(), s.client, tx)
	if err != nil {
		log.Printf("Error waiting for transaction %s: %v", tx.Hash().Hex(), err)
		s.db.UpdateWithdrawalStatus(withdrawal.ID, "failed", tx.Hash().Hex())
		return
	}

	if receipt.Status == 1 {
		s.db.UpdateWithdrawalStatus(withdrawal.ID, "success", tx.Hash().Hex())
		log.Printf("Withdrawal %d succeeded with tx %s", withdrawal.ID, tx.Hash().Hex())
	} else {
		s.db.UpdateWithdrawalStatus(withdrawal.ID, "failed", tx.Hash().Hex())
		switch withdrawal.TokenSymbol {
		case "BNB":
			s.db.UpdateAccountBalance(withdrawal.AccountID, "BNB", withdrawal.Amount)
		case "USDT":
			s.db.UpdateAccountBalance(withdrawal.AccountID, "USDT", withdrawal.Amount)
		}
		log.Printf("Withdrawal %d failed with tx %s, roll back account balance", withdrawal.ID, tx.Hash().Hex())
	}
}

func (s *WithdrawalService) encodeTransferCall(toAddress string, amount *big.Int) []byte {
	// This is a simplified version - in practice you'd use the ABI to properly encode
	// the function call: transfer(address,uint256)
	// For a real implementation, use the abi.JSON and abi.Pack methods
	data, err := s.usdtABI.Pack("transfer", common.HexToAddress(toAddress), amount)
	if err != nil {
		log.Printf("Failed to pack transfer call: %v", err)
		return []byte{}
	}
	return data
}
